src/avram/save_operation.cr

```diff
@@ -339,6 +339,37 @@ abstract class Avram::SaveOperation(T)
     !!record_id
   end
 
+  # `#persisted?` always returns `true` in `after_*` hooks, whether
+  # a new record was created, or an existing one was updated.
+  #
+  # This method should always return `true` for a create or `false`
+  # for an update, independent of the stage we are at in the operation.
+  def new_record? : Bool
+    {{ T.resolve.constant(:PRIMARY_KEY_NAME).id }}.value.nil?
+  end
+
+  def revert : self?
```

Comment:
should another value besides `nil` be returned?

Reply:
hmm... Yeah, I'm not sure about the `self?` return here 🤔  It's really returning either a new operation instance, or nothing as opposed to the current operation instance or nil. Maybe you want to know why a revert failed. We could have this either return a new operation if it succeeded, or return the old operation if it failed, and add to the `errors` object using `add_error(:revert, "Record must be saved before reverting")` or `add_error(:revert, "Could not delete record")`

Reply:
I like the idea of adding an `error` to the retruned record and returning the record.  it means you have to check the record for failures but it mean you can add more context to the reverted record.

Reply:
> Yeah, I'm not sure about the self?

I wasn't sure about it either. I just needed a way to represent a failed revert. `#add_error` seems a good fit, although I am not sure mutating the current operation is a good idea. if a revert failed, you may end up with `operation.saved?` being `true` while `operation.valid?` would be `false` for the same operation. I don't know if that would be OK?

> It's really returning either a new operation instance, or nothing as opposed to the current operation instance or nil

We already have the current operation, on which you are calling `#revert`, so returning the same operation may be redundant.

> I like the idea of adding an error to the retruned record and returning the record

Sounds good. It means you can call `operation.revert.valid?` to check if revert succeeded, while maintaining the state of the current operation.

Reply:
I dont know if this work can be done in a follow up PR but just as an idea we could mimic the pattern Operations have.

This is what a save operation is like.
```crystal
SaveUser.create(params) do |operation, user|
  if user # the user was saved
    html Users::ShowPage, user: user
  else
    html Users::NewPage, save_user: operation
  end
end
```

Some api for reverting
```crystal
MyModel.revert do |operation, user|
  if user # the user was reverted
    successful_reverted
  else
    unsuccessful_reverted
  end
end
```

This might be overly complex for revert.

Reply:
It seems impossible to call revert on models, since a model has no knowledge of it's previous state. An operation, on the other hand, does; it holds previous values in the form of `#original_value`s. Or maybe I'm missing the point?